### PR TITLE
Fix sysSubscribe races

### PR DESCRIPTION
**Pull request by kozlovic**
Made changes to processSub() to accept a subscription which allows
the icb callback to be set prior to add the subscription to the
account's sublist, which prevent races.
Fixed some other racy conditions, notably in addServiceImportSub()

Signed-off-by: Ivan Kozlovic <ivan@synadia.com>

### CODE REVIEW

File: server/client.go [hunk 1]
```diff
@@ -2123,7 +2123,10 @@ func (c *client) processSub(argo []byte, noForward bool) (*subscription, error)
 	default:
 		return nil, fmt.Errorf("processSub Parse Error: '%s'", arg)
 	}
+	return c.processSub(sub, noForward)
+}
 
+func (c *client) processSub(sub *subscription, noForward bool) (*subscription, error) {
```
Comment (kozlovic) [inline server/client.go:2129]: @derekcollison It seems that on Sublist.Insert() error, we send an error to the client but return `nil, nil` (no subscription, no error), which forces the callers to possibly check for nil sub. I wonder if I should not change that to return `nil, err` but maybe in parseSub() to restore old behavior (that is no error returned otherwise parser will close connection)?

File: server/accounts.go [hunk 2]
```diff
@@ -1355,43 +1355,47 @@ func (a *Account) subscribeInternal(subject string, cb msgHandler) (*subscriptio
...
 	}
-
-	sub.icb = func(sub *subscription, c *client, subject, reply string, msg []byte) {
-		c.processServiceImport(si, a, msg)
-	}
-
 	a.mu.Lock()
-	a.isid++
-	si.sub = sub
+	// TODO: Normally, we don't have to set si.sub here since it was done prior to
```
Comment (kozlovic) [inline server/accounts.go:1393]: @derekcollison Actually, since we would fail above if we detect that si.sub is not nil, this is likely a situation that will never occur, so I could remove this code. About err checking, see comment in processSub().

File: server/client.go (hunk 1 above)
Comment (kozlovic) [reply server/client.go:2132]: I have a second commit with this change and I think it is better.

File: server/accounts.go (hunk 2 above)
Comment (kozlovic) [reply server/accounts.go]: I have removed this part in the 2nd commit.

File: server/client.go (hunk 1 above)
Comment (derekcollison) [reply server/client.go:2132]: Unclear, I know we did not want to return an error for some reason.

File: server/accounts.go [hunk 3]
```diff
@@ -1355,43 +1355,47 @@ func (a *Account) subscribeInternal(subject string, cb msgHandler) (*subscriptio
...
 	if si.sub != nil {
+		a.mu.Unlock()
 		return fmt.Errorf("duplicate call to create subscription for service import")
 	}
+	a.isid++
+	sid := strconv.FormatUint(a.isid, 10)
+	cb := func(sub *subscription, c *client, subject, reply string, msg []byte) {
+		c.processServiceImport(si, a, msg)
+	}
+	si.sub = createSub(c, si.from, sid, cb)
```
Comment (derekcollison) [inline server/accounts.go:1379]: Should createSub hang off of client versus always passing in as first arg?

File: server/accounts.go (hunk 2 above)
Comment (derekcollison) [reply server/accounts.go]: ok

File: server/events.go [hunk 4]
```diff
@@ -1203,6 +1203,12 @@ func (s *Server) sendAuthErrorEvent(c *client) {
 // required to be copied.
 type msgHandler func(sub *subscription, client *client, subject, reply string, msg []byte)
 
+// Helper that creates a subscription object and initializes with given
+// properties suitable for system subscriptions.
+func createSub(c *client, subject, sid string, cb msgHandler) *subscription {
```
Comment (derekcollison) [inline server/events.go:1208]: Make method for client I think.

File: server/accounts.go (hunk 3 above)
Comment (kozlovic) [reply server/accounts.go:1379]: I did at first but thought it might be misleading. Wanted to be clear that this was just a helper instead of having &subscription{} everywhere. But I can certainly make it a `*client` receiver.

Comment (derekcollison) [reply server/accounts.go:1379]: Good point, could go either way.

Comment (kozlovic) [reply server/accounts.go:1379]: And to be clear: as of now it looks weird and really would be simpler to have `processSub(queue, sid string, cb msgHandler)` and create the subscription inside `processSub()`. And I could certainly do that. But in the mqtt branch I would have to add some more property to the subscription (namely `qos`). So again, could add `qos` as param to `processSub()`, or like I did in mqtt branch pass "subOptions" (or like here `*subscription` instead of introducing a new object type).
So to summarize, for main branch I could still solve the races but make it simple by just changing `processSub()` with some params and modify it only in the mqtt branch because of the need to have more properties in the `*subscription`. Wdyt?

Comment (derekcollison) [reply server/accounts.go:1379]: I like that idea, then as more progress is made on MQTT we can revisit.

Comment (kozlovic) [reply server/accounts.go]: OK, pushed forced the changes. For the service import si.sub race, I would have had to keep the way it was, but realized that we were only using `si.sub` to get the `si.sub.sid` so I replaced `si.sub` with `si.sid` :-)
